File: Add-Digits/AD.cpp
#include <iostream>

using namespace std;

int AD(int a){
  int num = a, sum;

  do {
    sum = 0;
    while(num>0) {
      sum += num%10;
      num /= 10;
    }
    num = sum;
  }while(sum > 10);
    
  return sum;
}

int main(){

  int sum = AD(38);
  cout << sum;
  return 0;
}

File: Add-Digits/AD.ml
let rec add_digits (num: int) : int =
  let sepa num = List.map (fun a -> Char.code a - 48)
                   (List.init
                      (String.length (string_of_int num))
                      (String.get (string_of_int num)))
  and sum :'int list -> int = List.fold_left (+) 0;
  in
  if num <0 then -1
  else if num >= 10 then add_digits (sum (sepa num))
  else num
;;
